<!DOCTYPE html>
<html lang="en">
<head>
	<title>Genomic Data Processing and Visualization in R</title>
	<meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
	<link rel="icon" href="pictures/remixicon_slideshow-3-line.svg" sizes="any" type="image/svg+xml">
	<link rel="stylesheet" href="node_modules/@highlightjs/cdn-assets/styles/github.min.css" type="text/css"/>
	<link rel="stylesheet" href="node_modules/katex/dist/katex.min.css" type="text/css"/>
	<link rel="stylesheet" href="styles/styles.css">
    <style>
        /* Customize the theme */

		.shower {
			--slide-ratio: calc(16 / 9);
			--color-key: #709d9b;
			--color-white: #eaeaea;
			--slide-notes-max-width: 80ch;
		}


		/* Explicit notes control labels */

		.control {
			--radius: 0.75em;
			--line-height: 1.5em;
		}

		.control input[type='checkbox'] + label::after {
			text-align: center;
			content: "N";
			color: var(--color-off);
		}

		.control input[type='checkbox']:checked + label::after {
			content: "Y";
			color: var(--color-on);
		}

		pre.dense code {
			font-size: 0.75rem;
			line-height: 1.75;
		}
    </style>
</head>
<body class="shower list">

	<header class="caption">
		<h1>Genomic Data Processing and Visualization in R</h1>
		<p><a href="https://blog.liang2.tw/">Liang-Bo Wang</a>, 2022-11-28.</p>
	</header>

	<menu class="control">
		<li class="show-notes">
			<input type="checkbox" id="show-notes-toggle">
			<label for="show-notes-toggle">Slide notes</label>
		</li>
	</menu>

	<!-- Cover Slide -->
	<section id="cover" class="slide clear">
		<h2 id="talk-header">Genomic Data Processing<br>and Visualization in R</h2>
		<p id="talk-info">
			Liang-Bo Wang<br>
			R-Ladies Taipei<br>
			2022-11-28
		</p>
		<p id="usage-instr">
			<kbd>Esc</kbd> or swipe up<br>with three fingers to exit<br>
			<kbd>←</kbd> <kbd>→</kbd> to navigate<br>
			<a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
				<img alt="Creative Commons Attribution 4.0 International License"
					src="pictures/cc-by-4.0.svg"/>
			</a>
		</p>
		<figure>
			<img src="pictures/andrew-ridley-jR4Zf-riEjI-unsplash.jpg" class="cover w" alt="">
			<figcaption class="copyright bottom white">
				Photo by <a href="https://unsplash.com/@aridley88">Andrew Ridley</a>
				on <a href="https://unsplash.com/photos/jR4Zf-riEjI">Unsplash</a>
			</figcaption>
		</figure>
		<style>
			#talk-header, #talk-info {
				left: var(--slide-left-side);
				right: var(--slide-right-side);
				text-align: center;
				position: absolute;
				color: var(--color-black);
				margin: 0;
			}
			#talk-header {
				font-size: 3.5rem;
				font-weight: 600;
				line-height: 1.25;
				top: 6rem;
				left: 0;
				right: 0;
				padding: 1.25em calc(var(--slide-left-side) / 2);
			}
			#talk-info {
				font-size: 1.5rem;
				line-height: 1.5;
				bottom: 5rem;
			}
			#cover .copyright {
				left: calc(var(--slide-left-side) / 2);
			}
			#usage-instr {
				color: var(--color-white);
				font-size: 1rem;
				line-height: 1.75;
				right: calc(var(--slide-right-side) / 2);
				bottom: 20px;
				position: absolute;
				text-align: right;
				margin: 0;
			}
			#usage-instr a[rel="license"] {
				background-image: none;
			}
			#usage-instr a[rel="license"] img {
				height: 1em;
			}
			#usage-instr kbd {
				color: var(--color-white);
				background-color: var(--color-key);
			}
		</style>
	</section>

	<section id="about-me" class="slide">
		<h2>About me</h2>
		<div class="two columns r33">
			<ul>
				<li>PhD in Computational and Systems Biology<br>
					from Washington University in St. Louis</li>
				<li>My thesis focuses on cancer genomics</li>
				<li>Worked in related consortia including
					<a href="https://proteomics.cancer.gov/programs/cptac">CPTAC</a>,
					<a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">TCGA</a>,
					and <a href="https://humantumoratlas.org/">HTAN</a>
					with thousands of samples
				</li>
				<li>Led the data analysis of the <a href="https://doi.org/10.1016/j.ccell.2021.01.006">CPTAC Glioblastoma study</a></li>
			</ul>
		</div>
		<img
			class="fill h place r"
			src="pictures/my_protrait.jpg"
			alt="My protrait">
	</section>

	<section id="talk-goal" class="slide">
		<h2>Today's goal</h2>
	</section>

	<div id="quick-intro-genome" class="slide tight">
		<h2>Background knowledge about genomics</h2>
		<div class="two columns r40">
			<ul>
				<li>Genome (DNA) = source code of our cells</li>
				<li>DNA encodes RNA encodes proteins; proteins enable certain functions</li>
				<li>Capture the activity of DNA/RNA/protein/... <br>
					= know what's going on in a cell</li>
				<li>High-throughput technologies help us capture cell-wide signals (DNA-seq, RNA-seq, microarray, mass spectrometry, ...)</li>
			</ul>
			<figure>
				<img class="fill w"
				src="pictures/wikimedia_chromosome.svg"
				alt="Graphic decomposition of a chromosome (found in the cell nucleus), to the bases pair of the DNA.">
				<figcaption class="citation right">
					<a href="https://commons.wikimedia.org/wiki/File:Chromosome_en.svg">Derivative work of KES47</a>,
					<a href="https://creativecommons.org/licenses/by/3.0">CC BY 3.0</a>,
					via Wikimedia Commons
				</figcaption>
			</figure>
		</div>
	</div>

	<section id="bioconductor" class="slide">
		<h2>R/Bioconductor is a powerful ecosystem for genomic analysis</h2>
		<ul>
			<li>Process and calculate genomic ranges</li>
			<li>Query annotations like genome sequences, genes, and microarray designs</li>
			<li>Store experiment results with annotations and metadata</li>
		</ul>
		<p>Plus R's strong data analysis foundation:<br>
			Data exploration/processing, visualization and statistics
		</p>
	</section>

	<section id="granges-cheatsheet" class="slide">
		<h2><code>GRanges</code> cheatsheet</h2>
	<pre class="language-text main-code">
<code>&gt; gr = exons(EnsDb.Hsapiens.v86); gr
GRanges object with 748400 ranges and 1 metadata column:
                  seqnames            ranges strand |         exon_id
                     &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
  ENSE00002234944        1       11869-12227      + | ENSE00002234944
  ENSE00001948541        1       12010-12057      + | ENSE00001948541
  ENSE00001671638        1       12179-12227      + | ENSE00001671638
              ...      ...               ...    ... .             ...
  ENSE00001638296        Y 26633345-26633431      - | ENSE00001638296
  ENSE00001797328        Y 26634523-26634652      - | ENSE00001797328
  ENSE00001794473        Y 56855244-56855488      + | ENSE00001794473
-------
seqinfo: 357 sequences (1 circular) from GRCh38 genome</code>
		</pre>
		<p class="citation">Adapted from <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan (2015)</p>
		<div class="next">
			<div class="highlight granges"></div>
			<div class="command granges">
				<em>GRanges</em>
				<pre class="text dense">
<code>length(gr)
gr[1:5]; names(gr)
seqnames(gr)
start(gr)
end(gr)
width(gr)
strand(gr)</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight dataframe"></div>
			<div class="command dataframe">
				<em>DataFrame</em>
				<pre class="text dense">
<code>mcols(gr)
gr$exon_id</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight seqinfo"></div>
			<div class="command seqinfo">
				<em>SeqInfo</em>
				<pre class="text dense">
<code>seqinfo(gr)
seqlevels(gr)
seqlengths(gr)
seqlevelsStyle(gr)</code>
				</pre>
			</div>
		</div>
		<style>
			#granges-cheatsheet {
				--color-granges: hsla(183, 87%, 43%, 0.15);
				--color-dataframe: hsla(35, 87%, 43%, 0.15);
				--color-seqinfo: hsla(208, 87%, 43%, 0.15);
			}
			#granges-cheatsheet .main-code :is(pre, code){
				font-size: 0.875rem;
				line-height: 1.5;
			}
			#granges-cheatsheet .citation {
				font-size: 0.8rem;
				line-height: 1rem;
			}
			#granges-cheatsheet .highlight {
				position: absolute;
			}
			#granges-cheatsheet .command {
				position: absolute;
				padding: 0 0.5em 0.5em 0.5em;
				font-size: 1rem;
				left: 750px;
				min-width: 10em;
			}
			#granges-cheatsheet .command pre {
				margin: 0 0 0 0.5em;
				padding: 0px;
			}
			#granges-cheatsheet .command pre code {
				width: 100%;
				line-height: 1.5;
				margin: 0px;
				padding: 0px;
			}
			#granges-cheatsheet .granges.highlight {
				left: 245px;
				top: 200px;
				width: 295px;
				height: 190px;
				background: var(--color-granges);
			}
			#granges-cheatsheet .granges.command {
				top: 120px;
				background-color: var(--color-granges);
			}

			#granges-cheatsheet .dataframe.highlight {
				left: 555px;
				top: 200px;
				width: 140px;
				height: 190px;
				background: var(--color-dataframe);
			}
			#granges-cheatsheet .dataframe.command {
				top: 295px;
				background-color: var(--color-dataframe);
			}

			#granges-cheatsheet .seqinfo.highlight {
				left: 92px;
				top: 407px;
				width: 473px;
				height: 27px;
				background: var(--color-seqinfo);
			}
			#granges-cheatsheet .seqinfo.command {
				top: 380px;
				background-color: var(--color-seqinfo);
			}
		</style>
	</section>

	<section id="grangeslist-cheatsheet" class="slide tight">
		<h2><code>GRangesList</code> cheatsheet</h2>
		<pre class="language-text main-code">
<code>&gt; grl = transcriptsBy(EnsDb.Hsapiens.v86, "gene"); grl
GRangesList object of length 63970:
$ENSG00000000003
GRanges object with 5 ranges and 2 metadata columns:
      seqnames              ranges strand |           tx_id        tx_biotype
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;       &lt;character&gt;
  [1]        X 100627109-100637104      - | ENST00000612152    protein_coding
  [2]        X 100632063-100637104      - | ENST00000614008    protein_coding
  [3]        X 100628670-100636806      - | ENST00000373020    protein_coding
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome

$ENSG00000000005
GRanges object with 2 ranges and 2 metadata columns:
      seqnames              ranges strand |           tx_id        tx_biotype
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;       &lt;character&gt;
  [1]        X 100584802-100599885      + | ENST00000373031    protein_coding
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome

...
&lt;63967 more elements></code>
		</pre>
		<p class="citation">Adapted from <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan (2015)</p>
		<div class="next">
			<div class="highlight grangeslist"></div>
			<div class="command grangeslist">
				<em>GRangesList</em>
				<pre class="language-text dense">
<code>(list of GRanges)
length(grl)
grl[1:4]
shift(grl, 1)
range(grl)
unlist(grl)</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight granges"></div>
			<div class="command granges">
				<em>GRanges</em>
				<pre class="language-text dense">
<code>grl[[2]]
grl[["ENSG00000000005"]]
grl[, "tx_biotype"]</code>
				</pre>
			</div>
		</div>
		<style>
			#grangeslist-cheatsheet {
				--color-grangeslist: hsla(260, 75%, 45%, 0.1);
				--color-granges: hsla(183, 87%, 43%, 0.15);
			}
			#grangeslist-cheatsheet .main-code :is(pre, code){
				font-size: 0.75rem;
				line-height: 1.5;
			}
			#grangeslist-cheatsheet .main-code {
				margin-bottom: 0.5em;
			}

			#grangeslist-cheatsheet .citation {
				font-size: 0.8rem;
				line-height: 1rem;
				text-align: right;
			}
			#grangeslist-cheatsheet .highlight {
				position: absolute;
			}
			#grangeslist-cheatsheet .command {
				position: absolute;
				padding: 0 0.5em 0.5em 0.5em;
				font-size: 1rem;
				left: 700px;
				min-width: 11em;
			}
			#grangeslist-cheatsheet .command pre {
				margin: 0 0 0 0.5em;
				padding: 0px;
			}
			#grangeslist-cheatsheet .command pre code {
				width: 100%;
				line-height: 1.5;
				margin: 0px;
				padding: 0px;
			}

			#grangeslist-cheatsheet .grangeslist.highlight {
				left: 90px;
				top: 145px;
				width: 580px;
				height: 380px;
				background: var(--color-grangeslist);
			}
			#grangeslist-cheatsheet .grangeslist.command {
				top: 145px;
				background-color: var(--color-grangeslist);
			}

			#grangeslist-cheatsheet .granges.highlight {
				left: 90px;
				top: 340px;
				width: 580px;
				height: 130px;
				background: var(--color-granges);
			}
			#grangeslist-cheatsheet .granges.command {
				top: 340px;
				background-color: var(--color-granges);
			}
		</style>
	</section>

	<section id="summarizedexperiment-cheatsheet" class="slide">
		<h2><code>SummarizedExperiment</code> cheatsheet</h2>
		<div class="columns two">
			<figure>
				<img class="fill w" src="pictures/summarizedexperiment_overview.svg" alt="">
				<figcaption class="citation">
					Ref: <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment's vignette</a>
				</figcaption>
			</figure>
			<ul>
				<li>A SummarizedExperiment object keeps the data and metadata (both samples and features) together</li>
				<li>Useful for organizing -omics data</li>
				<li>Export/Reload the object as a RDS file</li>
			</ul>
		</div>
		<style>
			#summarizedexperiment-cheatsheet h2 {
				font-size: 2.75rem;
			}
			#summarizedexperiment-cheatsheet ul {
				align-self: center;
				margin-bottom: 0;
			}
		</style>
	</section>

	<section id="tidyverse-integration" class="slide">
		<h2>Integration to tidyverse</h2>
		plyranges
	</section>

	<section id="purrr" class="slide">
		<h2>purrr</h2>
	</section>

	<section id="data-frame-comparison" class="slide">
		<h2>Data.frame, tibble, or DataFrame?</h2>
		as_tibble(rownames="my_row_name")
	</section>

	<section id="subset-or-filter" class="slide">
		<h2>filter vs subset</h2>
	</section>

	<section id="bioconductor-learning-resources" class="slide">
		<h2>How to learn Bioconductor?</h2>
	</section>


	<div id="x-inactivation" class="slide two columns r33">
		<div>
			<h2>X-inactivation</h2>
		</div>
		<figure>
			<img class="fill w"
				src="pictures/wikimedia_a_calico_cat.jpg"
				alt="A calico cat sitting on an indoor porch">
			<figcaption class="citation">
				<a href="https://commons.wikimedia.org/wiki/File:A_Calico_cat.jpg">Ellisn95</a>,
				<a href="https://creativecommons.org/licenses/by-sa/4.0">CC BY-SA 4.0</a>,
				via Wikimedia Commons
			</figcaption>
		</figure>
	</div>

	<div class="progress"></div>

	<!-- GitHub badge -->
	<footer class="badge">
		<a href="https://github.com/ccwang002/2022Talk-genomics-heatmap">
			<svg width="81" height="79" viewBox="0 0 81 79" aria-hidden="true">
				<path d="M40.5 0a40.5 40.5 0 0 0-12.8 78.93c2 .37 2.76-.88 2.76-2s0-3.51-.05-6.89c-11.27 2.45-13.64-5.43-13.64-5.43-1.84-4.68-4.5-5.92-4.5-5.92-3.68-2.51.28-2.46.28-2.46 4.06.29 6.2 4.17 6.2 4.17 3.61 6.19 9.48 4.4 11.79 3.37a8.65 8.65 0 0 1 2.57-5.41c-9-1-18.45-4.5-18.45-20a15.65 15.65 0 0 1 4.17-10.87 14.57 14.57 0 0 1 .4-10.72s3.4-1.09 11.14 4.15a38.39 38.39 0 0 1 20.28 0c7.73-5.24 11.13-4.15 11.13-4.15a14.55 14.55 0 0 1 .4 10.72 15.63 15.63 0 0 1 4.16 10.87c0 15.56-9.47 19-18.49 20 1.45 1.25 2.75 3.72 2.75 7.5v11.11c0 1.08.73 2.34 2.78 1.95A40.51 40.51 0 0 0 40.5 0z"/>
			</svg>
			See related materials<br>on GitHub
		</a>
	</footer>

	<script src="node_modules/@shower/core/dist/shower.js"></script>
	<script defer src="node_modules/@highlightjs/cdn-assets/highlight.min.js"></script>
	<script defer src="node_modules/katex/dist/katex.min.js"></script>
	<script defer src="node_modules/katex/dist/contrib/auto-render.min.js"></script>
	<script defer src="node_modules/katex/dist/contrib/copy-tex.min.js"></script>
	<script>
		document.addEventListener('DOMContentLoaded', (event) => {
			// Display or hide the slide notes
			const body = document.querySelector('body');
			const notesToggle = document.querySelector('.control .show-notes input');
			notesToggle.checked ? body.classList.add('show-notes') : body.classList.remove('show-notes');
			notesToggle.addEventListener('click', (event) => {
				body.classList.toggle('show-notes');
			});

			// Load Highlight.js
			hljs.highlightAll();

			// Load KaTeX
			renderMathInElement(document.body, {
				// Left aligned display equations
				// Details at https://github.com/KaTeX/KaTeX/issues/1990
				fleqn: true,
				delimiters: [
					{ left: '$$', right: '$$', display: true },
					{ left: '$', right: '$', display: false },
					{ left: '\\(', right: '\\)', display: false },
					{ left: '\\[', right: '\\]', display: true }
				]
			});
		});
	</script>
</body>
</html>
