<!DOCTYPE html>
<html lang="en">
<head>
	<title>Genomic Data Processing and Visualization in R</title>
	<meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
	<link rel="icon" href="pictures/remixicon_slideshow-3-line.svg" sizes="any" type="image/svg+xml">
	<link rel="stylesheet" href="node_modules/@highlightjs/cdn-assets/styles/github.min.css" type="text/css"/>
	<link rel="stylesheet" href="node_modules/katex/dist/katex.min.css" type="text/css"/>
	<link rel="stylesheet" href="styles/styles.css">
    <style>
        /* Customize the theme */

		.shower {
			--slide-ratio: calc(16 / 9);
			--color-key: #709d9b;
			--color-white: #eaeaea;
			--slide-notes-max-width: 80ch;
		}


		/* Explicit notes control labels */

		.control {
			--radius: 0.75em;
			--line-height: 1.5em;
		}

		.control input[type='checkbox'] + label::after {
			text-align: center;
			content: "N";
			color: var(--color-off);
		}

		.control input[type='checkbox']:checked + label::after {
			content: "Y";
			color: var(--color-on);
		}

		pre.dense code {
			font-size: 0.875rem;
			line-height: 1.75;
		}

		.slide pre code {
			font-weight: 450;
		}

		.place.top.right.biocsticker {
			width: 120px;
			top: 1rem;
			right: 1rem;
		}

		.slide .citation {
			font-size: 0.8rem;
			line-height: 1rem;
    		margin: 0;
		}

		.slide .citation.left {
 			text-align: left;
		}

		.slide .citation.right {
  	  		text-align: right;
		}

		.slide .citation.center {
    		text-align: center;
		}

    </style>
</head>
<body class="shower list">

	<header class="caption">
		<h1>Genomic Data Processing and Visualization in R</h1>
		<p><a href="https://blog.liang2.tw/">Liang-Bo Wang</a>, 2022-11-28.</p>
	</header>

	<menu class="control">
		<li class="show-notes">
			<input type="checkbox" id="show-notes-toggle">
			<label for="show-notes-toggle">Slide notes</label>
		</li>
	</menu>

	<!-- Cover Slide -->
	<section id="cover" class="slide clear">
		<h2 id="talk-header">Genomic Data Processing<br>and Visualization in R</h2>
		<p id="talk-info">
			Liang-Bo Wang<br>
			R-Ladies Taipei<br>
			2022-11-28
		</p>
		<p id="talk-details">
			<span class="web-instructions">
				<kbd>Esc</kbd> or swipe up<br>with three fingers to exit<br>
				<kbd>←</kbd> <kbd>→</kbd> to navigate<br>
			</span>
			<a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
				<img alt="Creative Commons Attribution 4.0 International License"
					src="pictures/cc-by-4.0.svg"/>
			</a>
		</p>
		<figure>
			<img src="pictures/andrew-ridley-jR4Zf-riEjI-unsplash.jpg" class="cover w" alt="">
			<figcaption class="copyright bottom white">
				Photo by <a href="https://unsplash.com/@aridley88">Andrew Ridley</a>
				on <a href="https://unsplash.com/photos/jR4Zf-riEjI">Unsplash</a>
			</figcaption>
		</figure>
		<style>
			#talk-header, #talk-info {
				left: var(--slide-left-side);
				right: var(--slide-right-side);
				text-align: center;
				position: absolute;
				color: var(--color-black);
				margin: 0;
			}
			#talk-header {
				font-size: 3.5rem;
				font-weight: 600;
				line-height: 1.25;
				top: 6rem;
				left: 0;
				right: 0;
				padding: 1.25em calc(var(--slide-left-side) / 2);
			}
			#talk-info {
				font-size: 1.5rem;
				line-height: 1.5;
				bottom: 5rem;
			}
			#cover .copyright {
				left: calc(var(--slide-left-side) / 2);
			}
			#talk-details {
				color: var(--color-white);
				font-size: 1rem;
				line-height: 1.75;
				right: calc(var(--slide-right-side) / 2);
				bottom: 20px;
				position: absolute;
				text-align: right;
				margin: 0;
			}
			#talk-details a[rel="license"] {
				background-image: none;
			}
			#talk-details a[rel="license"] img {
				height: 1em;
			}
			#talk-details kbd {
				color: var(--color-white);
				background-color: var(--color-key);
			}
			@media print {
				#cover .web-instructions {
					display: none;
				}
			}
		</style>
	</section>

	<section class="slide-notes" style="grid-row: span 3;">
		<h1>Slide notes</h1>
		<p><strong>This is not a tutorial.</strong>
			The goal of this talk is to introduce the related packages and provide some keywords for each task.
			Since it is going to cover <em>a lot</em> of different topics, this talk assumes the reader has the basic understanding of:</p>
		<ul>
			<li>R; familiar with the main topics mentioned in <a href="https://r4ds.had.co.nz/index.html">R for Data Science</a> (a great entry-level book)</li>
			<li>Pipe operators (<code>|></code> or <code>%>%</code>); see below</li>
			<li>tidyverse packages (readr, dplyr, purrr, ggplot2)</li>
			<li>grid graphics (mostly <code>gpar(...)</code> parameters)</li>
		</ul>
		<p>
			Since R 4.1, R introduces a native <a href="https://r4ds.had.co.nz/pipes.html">pipe</a> operator <code>|></code>.
			For example, the following code are equivalent:
		</p>
		<pre class="language-r">
<code>my_df = data.frame(col_A = 1:3)
# no pipe
transform(subset(my_df, col_A != 1), col_B = col_A + 3)
# native pipe in R 4.1+
my_df |>
	subset(col_A != 1) |>
	transform(col_B = col_A + 3)
# magrittr pipe
my_df %>%
	subset(col_A != 1) %>%
	transform(col_B = col_A + 3)</code>
		</pre>
	</section>

	<section id="how-to-draw-an-owl" class="slide">
		<img class="place center"
			src="pictures/how_to_draw_an_owl.jpg"
			alt="A meme showing how to draw an owl in two steps: 1. draw some circles. 2. draw the rest of the fucking owl.">
		<span class="goal next">Our goal</span>
		<style>
			#how-to-draw-an-owl .goal {
				position: absolute;
				bottom: 100px;
				left: 300px;
				text-align: center;
				font-size: 2rem;
				color: var(--color-key);
				font-weight: bold;
			}
		</style>
	</section>

	<section id="about-me" class="slide clear">
		<h2>About me</h2>
		<div class="two columns r33">
			<ul>
				<li>PhD in Computational and Systems Biology<br>
					from Washington University in St. Louis</li>
				<li>Thesis research: cancer proteogenomics (analysis and tool/pipeline development)</li>
				<li>Worked in related consortia including
					<a href="https://proteomics.cancer.gov/programs/cptac">CPTAC</a>,
					<a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">TCGA</a>,
					and <a href="https://humantumoratlas.org/">HTAN</a> with thousands of samples
				</li>
			</ul>
		</div>
		<img
			class="fill h place r"
			src="pictures/my_protrait.jpg"
			alt="My protrait">
	</section>

	<section id="quick-intro-genome" class="slide">
		<h2>Background knowledge about genomics</h2>
		<div class="two columns r40">
			<ul>
				<li>Genome (DNA) = source code of our cells</li>
				<li>If we measure DNA/RNA/protein/...,<br>
					we know what's going on in a cell (?)</li>
				<li>High-throughput technologies enable us to capture cell-wide/genome-wide signals<br>
					(DNA-seq, RNA-seq, microarray, mass spectrometry, imaging, ...)</li>
			</ul>
			<figure>
				<img class="fill w"
				src="pictures/wikimedia_chromosome.svg"
				alt="Graphic decomposition of a chromosome (found in the cell nucleus), to the bases pair of the DNA.">
				<figcaption class="citation right">
					<a href="https://commons.wikimedia.org/wiki/File:Chromosome_en.svg">KES47</a>,
					<a href="https://creativecommons.org/licenses/by/3.0">CC BY 3.0</a>,
					via Wikimedia Commons
				</figcaption>
			</figure>
		</div>
	</section>

	<section id="bioconductor" class="slide">
		<h2><strong style="font-weight: 400;">R/Bioconductor</strong>: a powerful ecosystem for genomic analysis</h2>
		<ul>
			<li>Process and calculate <strong>genomic ranges</strong></li>
			<li><strong>Query annotations</strong> like genome sequences, genes, and microarray designs</li>
			<li><strong>Store experiment results</strong> together with annotations and metadata</li>
			<li>Even better with R's strong data analysis foundation</li>
		</ul>
		<p>Install <a href="https://bioconductor.org/">Bioconductor</a> and its packages by <code>BiocManager::install()</code>
		</p>
		<img class="place top right biocsticker"
			src="pictures/Bioconductor_BiocSticker.png"
			alt="Hexagon sticker for Bioconductor">
	</section>

	<section id="outline" class="slide">
		<h2>Outline</h2>
		<div class="two columns l67">
			<ol>
				<li>Understand the basic data structures for genomics
					(<a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>: <code>GRanges</code> and <code>GRangesList</code>)</li>
				<li>Query genome annotations (<a href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a>)</li>
				<li>Store experiment results (<a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>)</li>
				<li>Plot heatmap visualizations (<a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap</a>)</li>
				<li>A demo for data exploration and analysis</li>
				<li>Challenges for future multi-omics data analysis</li>
			</ol>
			<figure>
				<img
				class="fill w"
				src="pictures/genomic_analysis_rocks_DALL·E.png"
				alt="Illustration of a rocket in the shape of a DNA helix traveling in space. DALL·E generated art.">
			</figure>
		</div>
		<style>
			#outline .columns {
				grid-column-gap: 1rem;
			}
			#outline code {
				font-size: 0.875em;
			}
			#outline figure {
				align-self: center;
			}
		</style>
	</section>

	<section id="genomicsranges-overview" class="slide">
		<h2><code>GenomicRanges</code> overview</h2>
		<ul>
			<li>How to specify a location on genome?
				A coorindate range of a chromosome on a genome reference,
				like chr1:100–150:- on hg38 (reverse strand)
			</li>
			<li>Coordinate system: 1-based, closed interval, strand aware, left-most (start <= end)</li>
			<li><a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> package provides <code>GRanges</code> and <code>GRangesList</code></li>
			<li>The building blocks for essentially all other genomic packages</li>
			<li>Check out <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">its vignettes</a> and <a href="https://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">this tutorial in 2015</a> for an in-depth walkthrough</li>
		</ul>
		<img class="place top right biocsticker"
			src="pictures/GenomicRanges_BiocSticker.png"
			alt="Hexagon sticker for the Bioconductor package GenomicRanges">
	</section>

	<section id="granges-cheatsheet" class="slide">
		<h2><code>GRanges</code> cheatsheet</h2>
		<pre class="language-text main-code">
<code>&gt; gr = exons(EnsDb.Hsapiens.v86); gr
GRanges object with 748400 ranges and 1 metadata column:
                  seqnames            ranges strand |         exon_id
                     &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;
  ENSE00002234944        1       11869-12227      + | ENSE00002234944
  ENSE00001948541        1       12010-12057      + | ENSE00001948541
  ENSE00001671638        1       12179-12227      + | ENSE00001671638
              ...      ...               ...    ... .             ...
  ENSE00001638296        Y 26633345-26633431      - | ENSE00001638296
  ENSE00001797328        Y 26634523-26634652      - | ENSE00001797328
  ENSE00001794473        Y 56855244-56855488      + | ENSE00001794473
-------
seqinfo: 357 sequences (1 circular) from GRCh38 genome</code>
		</pre>
		<p class="citation">Adapted from <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan (2015)</p>
		<div class="next">
			<div class="highlight granges"></div>
			<div class="command granges">
				<em>GRanges</em>
				<pre class="text dense">
<code>length(gr)
gr[1:5]; names(gr)
seqnames(gr)
start(gr)
end(gr)
width(gr)
strand(gr)</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight dataframe"></div>
			<div class="command dataframe">
				<em>DataFrame</em>
				<pre class="text dense">
<code>mcols(gr)
gr$exon_id</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight seqinfo"></div>
			<div class="command seqinfo">
				<em>Seqinfo</em>
				<pre class="text dense">
<code>seqinfo(gr)
seqlevels(gr)
seqlengths(gr)
seqlevelsStyle(gr)</code>
				</pre>
			</div>
		</div>
		<style>
			#granges-cheatsheet {
				--color-granges: hsla(183, 87%, 43%, 0.15);
				--color-dataframe: hsla(35, 87%, 43%, 0.15);
				--color-seqinfo: hsla(208, 87%, 43%, 0.15);
			}
			#granges-cheatsheet .main-code :is(pre, code){
				font-size: 0.875rem;
				line-height: 1.5;
			}
			#granges-cheatsheet .citation {
				font-size: 0.8rem;
				line-height: 1rem;
			}
			#granges-cheatsheet .highlight {
				position: absolute;
			}
			#granges-cheatsheet .command {
				position: absolute;
				padding: 0 0.5em 0.5em 0.5em;
				font-size: 1rem;
				left: 750px;
				min-width: 11em;
			}
			#granges-cheatsheet .command pre {
				margin: 0 0 0 0.5em;
				padding: 0px;
			}
			#granges-cheatsheet .command pre code {
				width: 100%;
				line-height: 1.5;
				margin: 0px;
				padding: 0px;
			}
			#granges-cheatsheet .granges.highlight {
				left: 245px;
				top: 200px;
				width: 295px;
				height: 190px;
				background: var(--color-granges);
			}
			#granges-cheatsheet .granges.command {
				top: 90px;
				background-color: var(--color-granges);
			}

			#granges-cheatsheet .dataframe.highlight {
				left: 555px;
				top: 200px;
				width: 140px;
				height: 190px;
				background: var(--color-dataframe);
			}
			#granges-cheatsheet .dataframe.command {
				top: 288px;
				background-color: var(--color-dataframe);
			}

			#granges-cheatsheet .seqinfo.highlight {
				left: 92px;
				top: 407px;
				width: 473px;
				height: 27px;
				background: var(--color-seqinfo);
			}
			#granges-cheatsheet .seqinfo.command {
				top: 380px;
				background-color: var(--color-seqinfo);
			}
		</style>
	</section>

	<section id="grangeslist-cheatsheet" class="slide tight">
		<h2><code>GRangesList</code> cheatsheet</h2>
		<pre class="language-text main-code">
<code>&gt; grl = transcriptsBy(EnsDb.Hsapiens.v86, "gene"); grl
GRangesList object of length 63970:
$ENSG00000000003
GRanges object with 5 ranges and 2 metadata columns:
      seqnames              ranges strand |           tx_id        tx_biotype
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;       &lt;character&gt;
  [1]        X 100627109-100637104      - | ENST00000612152    protein_coding
  [2]        X 100632063-100637104      - | ENST00000614008    protein_coding
  [3]        X 100628670-100636806      - | ENST00000373020    protein_coding
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome

$ENSG00000000005
GRanges object with 2 ranges and 2 metadata columns:
      seqnames              ranges strand |           tx_id        tx_biotype
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;       &lt;character&gt;
  [1]        X 100584802-100599885      + | ENST00000373031    protein_coding
  -------
  seqinfo: 357 sequences (1 circular) from GRCh38 genome

...
&lt;63967 more elements></code>
		</pre>
		<p class="citation">Adapted from <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan (2015)</p>
		<div class="next">
			<div class="highlight grangeslist"></div>
			<div class="command grangeslist">
				<em>GRangesList</em>
				<pre class="language-text dense">
<code>(list of GRanges)
length(grl)
grl[1:4]
shift(grl, 1)
range(grl)
unlist(grl)</code>
				</pre>
			</div>
		</div>
		<div class="next">
			<div class="highlight granges"></div>
			<div class="command granges">
				<em>GRanges</em>
				<pre class="language-text dense">
<code>grl[[2]]
grl[["ENSG00000000005"]]
grl[, "tx_biotype"]</code>
				</pre>
			</div>
		</div>
		<style>
			#grangeslist-cheatsheet {
				--color-grangeslist: hsla(260, 75%, 45%, 0.1);
				--color-granges: hsla(183, 87%, 43%, 0.15);
			}
			#grangeslist-cheatsheet .main-code :is(pre, code){
				font-size: 0.75rem;
				line-height: 1.5;
			}
			#grangeslist-cheatsheet .main-code {
				margin-bottom: 0.5em;
			}

			#grangeslist-cheatsheet .citation {
				font-size: 0.8rem;
				line-height: 1rem;
				text-align: right;
			}
			#grangeslist-cheatsheet .highlight {
				position: absolute;
			}
			#grangeslist-cheatsheet .command {
				position: absolute;
				padding: 0 0.5em 0.5em 0.5em;
				font-size: 1rem;
				left: 700px;
				min-width: 11em;
			}
			#grangeslist-cheatsheet .command pre {
				margin: 0 0 0 0.5em;
				padding: 0px;
			}
			#grangeslist-cheatsheet .command pre code {
				width: 100%;
				line-height: 1.5;
				margin: 0px;
				padding: 0px;
			}

			#grangeslist-cheatsheet .grangeslist.highlight {
				left: 90px;
				top: 145px;
				width: 580px;
				height: 380px;
				background: var(--color-grangeslist);
			}
			#grangeslist-cheatsheet .grangeslist.command {
				top: 145px;
				background-color: var(--color-grangeslist);
			}

			#grangeslist-cheatsheet .granges.highlight {
				left: 90px;
				top: 340px;
				width: 580px;
				height: 130px;
				background: var(--color-granges);
			}
			#grangeslist-cheatsheet .granges.command {
				top: 340px;
				background-color: var(--color-granges);
			}
		</style>
	</section>

	<section id="genomicranges-other-ways" class="slide">
		<h2>Other ways to construct a GRanges object</h2>
		<ul>
			<li>From a data frame: <code>makeGRangesFromDataFrame</code></li>
			<li>From external files (BED, GTF, and etc): use <a href="https://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a> package</li>
		</ul>
		<p>Good practice: always provide/verify the correct genome information<br>
			(<code>Seqinfo</code> object from <a href="https://bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html">GenomeInfoDb</a> package)</p>
	</section>

	<section id="genomicranges-methods" class="slide tight">
		<h2>GenomicRanges provides many methods to manipulate the ranges</h2>
		<pre class="language-r dense">
<code># All methods are strand aware
?"intra-range-methods"    # Methods applied to every range independently
shift(), narrow(), flank(), promoters(), resize(), ...

?"inter-range-methods"    # Methods applied across all ranges at once
range(), reduce(), gaps(), disjoin(), coverage(), ...

?"findOverlaps-methods"   # Find overlaps between two ranges
findOverlaps(), countOverlaps(), %over%, %within%, %outside%, ...

methods(class="GRanges")  # List all available methods</code>
		</pre>
		<p>More on <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan (2015)</p>
		<style>
			#genomicranges-methods pre.dense code {
				font-size: 1rem;
				line-height: 1.75;
			}
			#genomicranges-methods p {
				font-size: 1rem;
				text-align: right;
			}
		</style>
	</section>

	<section class="slide-notes">
		GenomeInfoDb is a powerful package for genome information

		<a href="https://blog.liang2.tw/posts/2019/02/gdc-seqinfo/">my blog post</a> to create a Seqinfo object from a genome <code>.dict</code> file.
	</section>

	<section id="ensembldb" class="slide">
		<h2>Query genome annotations by <code>ensembldb</code></h2>
		<ul>
			<li>Find the location of a genomic feature</li>
			<li>ID conversion (gene ↔ transcript ↔ protein)</li>
			<li>Fix a certain version of the annotation for a project</li>
			<li>
				<a href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a>
				provides a local copy of <a href="https://ensembl.org/">Ensembl</a> genome annotations (EnsDb.*)
			</li>
			<li>
				All EnsDb objects (different versions and species) are available on
				<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a>.
				For example, Ensembl release 102 (GENCODE 36) for human is available as <a href="https://annotationhub.bioconductor.org/ahid/AH89180"><code>AH89180</code></a> on AnnotationHub.
			</li>
		</ul>
		<img class="place top right biocsticker"
			src="pictures/ensembldb_BiocSticker.png"
			alt="Hexagon sticker for the Bioconductor package ensembldb">
		<style>
			#ensembldb h2 {
				font-size: 2.5rem;
			}
		</style>
	</section>

	<section id="ensembldb-usage" class="slide tight">
		<p>Example: Find all the protein coding genes in autosomes</p>
		<pre class="language-r dense">
<code>library(EnsDb.Hsapiens.v86)
my_filter = AnnotationFilter(~ gene_biotype == "protein_coding" & seq_name %in% 1:22)
genes(EnsDb.Hsapiens.v86, filter = my_filter) |>
    sortSeqlevels() |>  # Sort the chromosomes in natural order (1...22, X, Y, MT)
    sort(ignore.strand=TRUE)  # Sort the genes by on their genomic locations
# GRanges object with 19025 ranges and 2 metadata columns:
#                   seqnames            ranges strand |   gene_name   gene_biotype
#                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt;
#   ENSG00000186092        1       69091-70008      + |       OR4F5 protein_coding
#   ENSG00000279928        1     182393-184158      + |  FO538757.2 protein_coding
#   ENSG00000279457        1     184923-200322      - |  FO538757.1 protein_coding
#               ...      ...               ...    ... .         ...            ...
#   ENSG00000251322       22 50674415-50733298      + |      SHANK3 protein_coding
#   ENSG00000100312       22 50738196-50745334      + |         ACR protein_coding
#   ENSG00000079974       22 50767501-50783663      - |      RABL2B protein_coding
#   -------
#   seqinfo: 22 sequences from GRCh38 genome</code>
		</pre>
		<style>
			#ensembldb-usage p:first-of-type {
				margin-bottom: 0.5em;
			}
			#ensembldb-usage pre code {
				font-size: 0.875rem;
			}
		</style>
	</section>

	<section id="other-annotations" class="slide tight">
		<h2>Other annotations in R/Bioconductor</h2>
		<table class="striped">
		<thead>
			<tr>
				<th>Object type</th>
				<th>Example packages</th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<th role="row">EnsDb</th>
				<td>
					<a href="https://bioconductor.org/packages/EnsDb.Hsapiens.v86/">EnsDb.Hsapiens.v86</a>
					(TxDb compatible)
				</td>
			</tr>
			<tr>
				<th role="row">TxDb</th>
				<td>
					<a href="http://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene/">TxDb.Hsapiens.UCSC.hg38.knownGene</a>
				</td>
			</tr>
			<tr>
				<th role="row">BSgenome</th>
				<td>
					<a href="http://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg38/">BSgenome.Hsapiens.UCSC.hg38</a>
					(genome sequence)
				</td>
			</tr>
			<tr>
				<th role="row"><em>Others</em></th>
				<td>
					<a href="http://bioconductor.org/packages/org.Hs.eg.db/">org.Hs.eg.db</a> (organism),
					<a href="http://bioconductor.org/packages/GO.db/">GO.db</a> (system biology), ...
				</td>
			</tr>
		</tbody>
		</table>
		<p>
			See also:
			<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotationDbi</a>,
			<a href="https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a>,
			<a href="http://bioconductor.org/packages/biomaRt/">biomaRt</a> (web based),
			<a href="http://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a>,
			and other packages under the <a href="https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData">AnnotationData</a> category.
		</p>
		<p class="citation right">
			Adapted from Bioconductor's annotation workflow: <a href="https://bioconductor.org/packages/release/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">Genomic Annotation Resources</a>
		</p>
	</section>

	<section id="summarizedexperiment-cheatsheet" class="slide">
		<h2><code>SummarizedExperiment</code> cheatsheet</h2>
		<div class="columns two">
			<div>
				<ul>
					<li>An object to keep the data of a cohort and the related metadata (both samples and features) together</li>
					<li>Useful for organizing -omics data</li>
					<li>Export/Reload the object as a RDS file</li>
				</ul>
				<pre class="language-r dense">
<code>se = SummarizedExperiment(...)
se |> saveRDS("my_se.rds")
se = readRDS("my_se.rds")
</code>
				</pre>
			</div>
			<figure>
				<img class="fill w" src="pictures/summarizedexperiment_overview.svg" alt="">
				<figcaption class="citation right">
					Source: <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment's vignette</a>
				</figcaption>
			</figure>
		</div>
		<img class="place top right biocsticker"
			src="pictures/SummarizedExperiment_BiocSticker.png"
			alt="Hexagon sticker for the Bioconductor package SummarizedExperiment">
		<style>
			#summarizedexperiment-cheatsheet h2 {
				font-size: 2.5rem;
			}
			#summarizedexperiment-cheatsheet ul {
				margin-bottom: 0;
			}
		</style>
	</section>

	<section id="bioconductor-plus-tidyverse" class="slide">
	<h2>Bioconductor + tidyverse</h2>
		<ul>
			<li>I find <a href="https://www.tidyverse.org/">tidyverse</a>'s verbs easy to learn and use, personally</li>
			<li>Bioconductor and tidyverse are two <em>separate</em> ecosystems</li>
			<li>Function overrides are common (e.g. <code>select()</code>, <code>filter()</code>)</li>
			<li><code>DataFrame</code> (Bioconductor), <code>tibble</code> (tidyverse), <code>data.frame</code> are all different classes with slightly different behaviors (e.g., row names)</li>
			<li>Packages like <a href="https://sa-lee.github.io/plyranges/index.html">plyranges</a> and <a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> introduces "verbs" to Bioconductor classes</li>
		</ul>
		<img class="place top right biocsticker"
			src="pictures/tidyverse_sticker.png"
			alt="Hexagon sticker for tidyverse">
	</section>

	<section class="slide-notes" style="grid-row: span 2;">
		<h3>Function overrides</h3>
		<p>
			By default, the import order of the packages leads to different function overrides. Specify the function explicitly, say,
			<code>dplyr::select()</code> and <code>AnnotationDbi::filter()</code>.
		</p>
		<p>
			Alternatively, use packages like <a href="https://conflicted.r-lib.org/">conflicted</a> to explicitly resolve function conflicts.
		</p>

		<h3>Conversion between DataFrame and tibble</h3>
		<p>One of the main differences is the use of row names. Row names of DataFrame (and other Bioconductor packages) are very useful, while tibble will remove the row names after subsetting and discourage users to use them.</p>
		<pre class="language-r">
<code># Tibble to DataFrame (and use one column as rownames)
my_DF = my_tbl |>
	column_to_rownames("rowname_col") |>
	as("DataFrame")
# DataFrame to tibble (and extract rownames to one column)
my_tbl = my_DF |>
	as_tibble(rownames = "rowname_col")</code></pre>
	</section>

	<section id="purrr" class="slide">
		<h2>purrr's map functions are handy to process many samples in the same way</h2>
		<ul>
			<li>Usecases: read the specific format of a pipeline output</li>
			<li><a href="https://purrr.tidyverse.org/">purrr</a> provides a consistent interface to eliminate most for-loops:
				<ul>
					<li><code>my_list |> map(my_fun)</code>: the most basic form</li>
					<li><code>my_list |> imap(...)</code>:
						also pass the list key/name to the function</li>
					<li>
						<code>my_df |> pmap(function(col_A, col_B, ...) ...)</code>:<br>
						map a data frame of parameters as input
					</li>
				</ul>
			</li>
		</ul>
		<img class="place top right biocsticker"
			src="pictures/purrr_sticker.png"
			alt="Hexagon sticker for the purrr package">
	</section>

	<section id="bioconductor-learning-resources" class="slide">
		<h2>Ways to learn Bioconductor</h2>
		<ul>
			<li><a href="https://www.bioconductor.org/help/package-vignettes/">Package vignettes</a> are great to get started with a new package</li>
			<li>Bioconductor <a href="https://www.bioconductor.org/packages/release/workflows/">workflows</a> and <a href="https://www.bioconductor.org/help/course-materials/">courses</a> integrate multiple packages over a topic, such as <a href="https://www.bioconductor.org/packages/release/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">annotation</a></li>
			<li><a href="https://genomicsclass.github.io/book/">PH525x series - Biomedical Data Science</a> by Rafael Irizarry and Michael Love</li>
			<li><a href="https://www.huber.embl.de/msmb/">Modern Statistics for Modern Biology</a> by Susan Holmes and Wolfgang Huber</li>
			<li><a href="https://blog.liang2.tw/tag/bioconductor/">My blog posts</a> about my usage and experience of R/Bioconductor</li>
		</ul>
	</section>

	<section id="genomics-data-vis" class="slide">
		<ul>
			<li>Gviz, ggbio</li>
			<li>heatmap</li>
			<li>
		circos plot
			</li>
			<li>
		rasterization
			</li>
		</ul>
	</section>

	<section id="complexheatmap" class="slide tight">
		<h2>ComplexHeatmap: the ultimate way to make heatmaps</h2>
		<div class="two columns">
			<ul>
				<li>Heatmap is a basic way to dump high dimensional data in one plot and identify large trends</li>
				<li><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap</a> is <strong>highly configurable</strong></li>
				<li>Based on <a href="https://stat.ethz.ch/R-manual/R-devel/library/grid/html/grid-package.html">grid</a> graphics (check out <a href="https://www.stat.auckland.ac.nz/~paul/RG3e/">R Graphics, 3rd</a>)</li>
				<li>User directly controls the underlying parameters by <code>gpar(...)</code> (unlike <a href="https://ggplot2.tidyverse.org/">ggplot2</a>)</li>
				<li>We'll go through some examples from its <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">guidebook</a></li>
			</ul>
		<figure>
			<img class="fill w"
					src="pictures/complexheatmap_book_cover.jpg"
					alt="Cover of the ComplexHeatmap complete reference book">
			<figcaption class="citation">
				<a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">ComplexHeatmap Complete Reference</a>
				is an extensive guide to the package's design and all of the configurations.
			</figcaption>
		</figure>
		</div>
		<style>
			#complexheatmap .two.columns {
				grid-template-columns: 7fr 3fr;
			}
			#complexheatmap figure img {
				border: 2px solid var(--color-key);
			}
		</style>
	</section>

	<section id="hello-heatmap" class="slide tight">
		<h2>Hello, heatmap</h2>
		<div class="two columns">
			<pre class="language-r dense">
<code>mat = matrix(..., nr = 18, nc = 24)  # see notes

# 1. Define a color mapping function
col_fun = circlize::colorRamp2(
	c(-2, 0, 2), c("blue", "white", "red"))
c(0, 1, 2, 3) |> set_names() |> col_fun()
# "#FFFFFFFF" "#FF9E81FF" "#FF0000FF" "#FF0000FF"

# 2. Construct the heatmap
ht = Heatmap(
    mat, name = "mat",
    col = col_fun,
    na_col = "black"  # color for NAs
)
draw(ht)</code>
			</pre>
			<figure>
				<img class="fill w"
					src="pictures/complexheatmap_demo_1.png"
					alt="The first heatmap example generated by ComplexHeatmap">
			</figure>
		</div>
	</section>

	<section class="slide-notes" style="grid-row: span 3;">
		<h3>Code to generate the example matrix</h3>
		<p>The code was copied from the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html">first example</a> in the ComplexHeatmap Complete Reference (Chatper 2: A Single Heatmap).</p>
		<pre class="language-r">
<code>set.seed(123)
nr1 = 4; nr2 = 8; nr3 = 6; nr = nr1 + nr2 + nr3
nc1 = 6; nc2 = 8; nc3 = 10; nc = nc1 + nc2 + nc3
mat = cbind(
  rbind(
	matrix(rnorm(nr1*nc1, mean = 1, sd = 0.5), nr = nr1),
    matrix(rnorm(nr2*nc1, mean = 0, sd = 0.5), nr = nr2),
  	matrix(rnorm(nr3*nc1, mean = 0, sd = 0.5), nr = nr3)),
  rbind(
	matrix(rnorm(nr1*nc2, mean = 0, sd = 0.5), nr = nr1),
    matrix(rnorm(nr2*nc2, mean = 1, sd = 0.5), nr = nr2),
	matrix(rnorm(nr3*nc2, mean = 0, sd = 0.5), nr = nr3)),
  rbind(
	matrix(rnorm(nr1*nc3, mean = 0.5, sd = 0.5), nr = nr1),
    matrix(rnorm(nr2*nc3, mean = 0.5, sd = 0.5), nr = nr2),
    matrix(rnorm(nr3*nc3, mean = 1, sd = 0.5), nr = nr3))
)
# random shuffle rows and columns
mat = mat[sample(nr, nr), sample(nc, nc)]
# Add column and row names
rownames(mat) = paste0("row", seq_len(nr))
colnames(mat) = paste0("column", seq_len(nc))
# Add some missing values
mat[2:3, 2] = NA_real_</code>
		</pre>
	</section>

	<section id="heatmap-annotations" class="slide tight">
		<h2>Flexible heatmap annotations</h2>
		<div class="three columns">
			<pre class="language-r dense">
<code>my_colors = list(
    foo1 = hcl.colors(
		4, palette = "viridis"
	) |> set_names(LETTERS[1:4]),
    foo2 = circlize::colorRamp2(
        c(0, 1),
        hcl_palette = "blues"
    )
)
col_ha = columnAnnotation(
    foo1 = sample(
		LETTERS[1:4], 24,
	    replace = TRUE
	),
    foo2 = runif(24) |> sort(),
    foo3 = anno_barplot(runif(24)),
    col = my_colors
)
</code>
			</pre>
			<pre class="language-r dense next">
<code>row_ha = rowAnnotation(
    bar = anno_mark(
        at = c(7, 10, 18, 8, 1),
        labels = LETTERS[22:27],
        side = "left",
        labels_gp = gpar(fontsize=18)
))
ht = Heatmap(
	mat, name = "mat",
    col = col_fun,
	na_col = "black",
	# cut the dendrogram into 3
    column_split = 3,
    show_row_dend = FALSE,
    top_annotation = col_ha,
    left_annotation = row_ha,
)
draw(ht, merge_legends = TRUE)</code>
			</pre>
			<figure>
				<img class="fill w"
					src="pictures/complexheatmap_demo_2.png"
					alt="The second heatmap example generated by ComplexHeatmap, which added row and column annotations and split the heatmap columns into groups based on the column dendogram">
				<figcaption class="citation">Detail usage of annotations is available in the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/heatmap-annotations.html">guidebook</a></figcaption>
			</figure>
		</div>
		<style>
			#heatmap-annotations {
				--slide-right-side: 16px;
			}
			#heatmap-annotations .columns {
				column-gap: 1rem;
				grid-template-columns: 1fr 1fr 350px;
			}
			#heatmap-annotations pre code {
				font-size: 0.75rem;
				width: 100%;
			}
		</style>
	</section>

	<section id="heatmap-list" class="slide tight">
		<h2>Draw list of heatmaps and annotations</h2>
		<div class="two columns r40">
			<div>
				<ul>
					<li>Useful for combining heatmaps of different data types of the same sets of samples</li>
					<li>Match the row/column names and orders</li>
				</ul>
				<pre class="language-r dense">
<code># To concatenate heatmaps row-wise
mat1 = mat1[sample_order, ]
mat2 = mat2[sample_order, ]
anno_df = anno_df[sample_order, ]
ht_list = ht1(mat1) + ht2(mat2) + row_ha(anno_df)

# ... column-wise
ht_list = ht1 %v% ht2 %v% col_ha</code>
				</pre>
			</div>
			<figure>
				<img class="fill w"
					src="pictures/complexheatmap_design.png"
					alt="ComplexHeatmap design allows for list of heatmaps and annotations">
				<figcaption class="citation">
					Lists of heatmaps and annotations can be concatenated
				</figcaption>
			</figure>
		</div>
	</section>

	<section class="slide">
		<h2 class="shout">Demo on an example dataset</h2>
	</section>

	<section id="cptac-gbm-cohort" class="slide tight">
		<h2>CPTAC glioblastoma 2021 cohort</h2>
		<figure>
			<img class="fill w"
				src="pictures/cptac_gbm_2021_cohort_design.png"
				alt="Overview of the CPTAC glioblastoma 2021 discovery cohort design and the available data types">
			<figcaption class="citation right">
				Data publicly availabile at <abbr title="National Cancer Institute">NCI</abbr>'s <a href="https://gdc.cancer.gov/">Genomic Data Commons (GDC)</a>, <a href="https://pdc.cancer.gov/pdc/">Proteomic Data Commons (PDC)</a>, and <a href="https://www.cancerimagingarchive.net/">Cancer Imaging Archive (TCIA)</a><br>
				Wang <em>et al.</em> (2021). DOI: <a href="https://pubmed.ncbi.nlm.nih.gov/33577785/">10.1016/j.ccell.2021.01.006</a></figcaption>
		</figure>
		<div class="description">
			<ul>
				<li><a href="https://proteomics.cancer.gov/programs/cptac">CPTAC</a> prospective collection of treatment naïve glioblastoma (GBM) tumors (n=99)</li>
				<li><a href="https://www.gtexportal.org/">GTEx</a> brain (frontal cortex) normals (n=10; unmatched)</li>
			</ul>
		</div>
		<style>
			#cptac-gbm-cohort .description {
				position: absolute;
				width: 15rem;
				bottom: 100px;
				left: var(--slide-left-side);

				font-size: 0.875rem;
				line-height: 1.5;
			}
			#cptac-gbm-cohort .citation {
				line-height: 1.5;
			}
		</style>
	</section>

	<section id="x-inactivation" class="slide two columns r33">
		<div>
			<h2>X-inactivation</h2>
		</div>
		<figure>
			<img class="fill w"
				src="pictures/wikimedia_a_calico_cat.jpg"
				alt="A calico cat sitting on an indoor porch">
			<figcaption class="citation">
				<a href="https://commons.wikimedia.org/wiki/File:A_Calico_cat.jpg">Ellisn95</a>,
				<a href="https://creativecommons.org/licenses/by-sa/4.0">CC BY-SA 4.0</a>,
				via Wikimedia Commons
			</figcaption>
		</figure>
	</section>

	<section id="data-overview-final" class="slide">
		<div class="two columns">
			<div>
				<p>Final data overview figure in the <a href="https://doi.org/10.1016/j.ccell.2021.01.006">paper</a></p>
			</div>
			<div></div>
		</div>
		<img class="final-figure1 place right"
			src="pictures/cptac_gbm_figure1_data_overview.png"
			alt="Data overview figure of the cptac glioblastoma cohort in the final publication in 2021">
		<style>
			#data-overview-final .final-figure1 {
				height: 450px;
			}
		</style>
	</section>

	<div class="progress"></div>

	<!-- GitHub badge -->
	<footer class="badge">
		<a href="https://github.com/ccwang002/2022Talk-genomics-heatmap">
			<svg width="81" height="79" viewBox="0 0 81 79" aria-hidden="true">
				<path d="M40.5 0a40.5 40.5 0 0 0-12.8 78.93c2 .37 2.76-.88 2.76-2s0-3.51-.05-6.89c-11.27 2.45-13.64-5.43-13.64-5.43-1.84-4.68-4.5-5.92-4.5-5.92-3.68-2.51.28-2.46.28-2.46 4.06.29 6.2 4.17 6.2 4.17 3.61 6.19 9.48 4.4 11.79 3.37a8.65 8.65 0 0 1 2.57-5.41c-9-1-18.45-4.5-18.45-20a15.65 15.65 0 0 1 4.17-10.87 14.57 14.57 0 0 1 .4-10.72s3.4-1.09 11.14 4.15a38.39 38.39 0 0 1 20.28 0c7.73-5.24 11.13-4.15 11.13-4.15a14.55 14.55 0 0 1 .4 10.72 15.63 15.63 0 0 1 4.16 10.87c0 15.56-9.47 19-18.49 20 1.45 1.25 2.75 3.72 2.75 7.5v11.11c0 1.08.73 2.34 2.78 1.95A40.51 40.51 0 0 0 40.5 0z"/>
			</svg>
			See related materials<br>on GitHub
		</a>
	</footer>

	<script src="node_modules/@shower/core/dist/shower.js"></script>
	<script defer src="node_modules/@highlightjs/cdn-assets/highlight.min.js"></script>
	<script defer src="node_modules/katex/dist/katex.min.js"></script>
	<script defer src="node_modules/katex/dist/contrib/auto-render.min.js"></script>
	<script defer src="node_modules/katex/dist/contrib/copy-tex.min.js"></script>
	<script>
		document.addEventListener('DOMContentLoaded', (event) => {
			// Display or hide the slide notes
			const body = document.querySelector('body');
			const notesToggle = document.querySelector('.control .show-notes input');
			notesToggle.checked ? body.classList.add('show-notes') : body.classList.remove('show-notes');
			notesToggle.addEventListener('click', (event) => {
				body.classList.toggle('show-notes');
			});

			// Load Highlight.js
			hljs.highlightAll();

			// Load KaTeX
			renderMathInElement(document.body, {
				// Left aligned display equations
				// Details at https://github.com/KaTeX/KaTeX/issues/1990
				fleqn: true,
				delimiters: [
					{ left: '$$', right: '$$', display: true },
					{ left: '$', right: '$', display: false },
					{ left: '\\(', right: '\\)', display: false },
					{ left: '\\[', right: '\\]', display: true }
				]
			});
		});
	</script>
</body>
</html>
